aws/hll/aws-s3-transfer-manager/src/io/part_reader.rs

```diff
@@ -0,0 +1,336 @@
+    fn with_offset(self, offset: u64) -> Self {
+        Self { offset, ..self }
+    }
+}
+
+/// [ReadPart] implementation for in-memory input streams.
+#[derive(Debug)]
+struct BytesPartReader {
+    buf: Bytes,
+    part_size: usize,
+    state: Mutex<PartReaderState>,
+}
+
+impl BytesPartReader {
+    fn new(buf: Bytes, part_size: usize) -> Self {
+        let content_length = buf.remaining() as u64;
+        Self {
+            buf,
+            part_size,
+            state: Mutex::new(PartReaderState::new(content_length)),
+        }
+    }
+}
+
+impl ReadPart for BytesPartReader {
+    async fn next_part(&self) -> Result<Option<PartData>, Error> {
+        let mut state = self.state.lock().expect("lock valid");
+        if state.remaining == 0 {
+            return Ok(None);
+        }
+
+        let start = state.offset as usize;
+        let end = cmp::min(start + self.part_size, self.buf.len());
+        let data = self.buf.slice(start..end);
+        let part_number = state.part_number;
+        state.part_number += 1;
+        state.offset += data.len() as u64;
+        state.remaining -= data.len() as u64;
+        let part = PartData { data, part_number };
+        Ok(Some(part))
+    }
+}
+
+/// [ReadPart] implementation for path based input streams
+#[derive(Debug)]
+struct PathBodyPartReader {
+    body: PathBody,
+    part_size: usize,
+    state: Mutex<PartReaderState>,
+}
+
+impl PathBodyPartReader {
+    fn new(body: PathBody, part_size: usize) -> Self {
+        let offset = body.offset;
+        let content_length = body.length;
+        Self {
+            body,
+            part_size,
+            state: Mutex::new(PartReaderState::new(content_length).with_offset(offset)),
+        }
+    }
+}
+
+impl ReadPart for PathBodyPartReader {
+    async fn next_part(&self) -> Result<Option<PartData>, Error> {
+        let (offset, part_number, part_size) = {
+            let mut state = self.state.lock().expect("lock valid");
+            if state.remaining == 0 {
+                return Ok(None);
+            }
+            let offset = state.offset;
+            let part_number = state.part_number;
+
+            let part_size = cmp::min(self.part_size as u64, state.remaining);
+            state.offset += part_size;
+            state.part_number += 1;
+            state.remaining -= part_size;
+
+            (offset, part_number, part_size)
+        };
+        let path = self.body.path.clone();
+        let handle = tokio::task::spawn_blocking(move || {
+            // TODO(aws-sdk-rust#1159) - replace allocation with memory pool
+            let mut dst = BytesMut::with_capacity(part_size as usize);
+            // we need to set the length so that the raw &[u8] slice has the correct
+            // size, we are guaranteed to read exactly part_size data from file on success
+            // FIXME(aws-sdk-rust#1159) - can we get rid of this use of unsafe?
+            unsafe { dst.set_len(dst.capacity()) }
```

Comment:
I feel like there should be a `min` in here somewhere
```suggestion
            unsafe { dst.set_len(dst.len().min(dst.capacity())) }
```
but IDK. As long as we test the three cases (>, ==, and <)

Reply:
`dst.len()` is 0 after calling `with_capacity()`, that was the original issue here is that we are allocating a chunk and trying to fill the underlying slice but that slice is of length zero to start with because nothing has been written to `BytesMut` yet.